### DISCUSSION TITLE
Is --drop intended to only work with builtins? --drop does not remove a function I created myself.

### DISCUSSION THREAD

**Discussion by fredrik-hjarner** [oven-sh/bun#14960] [Q&A]
Regarding --drop paramater. Documentation:

> You can use --drop to remove function calls from your JavaScript bundle. This is useful if you want to remove debug code from your production bundle. For example, if you pass --drop=console, all calls to console.log will be removed from your code.

I was experimenting with --drop trying to remove a "debug-only" function, but noticed that "custom" functions will not be removed.

Is --drop intended to work only with/for builtins?

index.ts:

```
function test() {
    throw new Error("test");
}
test();
console.log("Hello via Bun!");
Bun.sleep(1000);
window.alert("Hello via Bun!");
```

cmd:
`bun build index.ts --drop="console" --drop="test" --drop="Bun" --drop="window"`

output:
```
// index.ts
function test() {
  throw new Error("test");
}
test();
```

**Comment by ericchase** [comment]
i too would like to know. i've tested this thoroughly and haven't found a way to get bun to drop anything that is actually defined, which seems to make the feature only marginally useful

**Comment by anachrohack** [comment]
Digging this up with some recent findings, though no solutions as of yet. the --drop command seems to only target things defined by `globalThis`:

source:
```ts
function test() {
    throw new Error("test");
}

const _x = {
    test
};
globalThis.x = _x;

test();
_x.test();
x.test();
```
command:

```sh
bun build test.ts --drop="x" --drop="_x" --drop="test"
```

output:
```js
// test.ts
function test() {
  throw new Error("test");
}
var _x = {
  test
};
globalThis.x = _x;
test();
_x.test();
```

Notice there's no `x.test()` at the bottom.

**Reply by ericchase** [reply]
massive find! would probably be a pain to use this effectively in typescript, but being able to drop from global is a significant improvement over just bun's built-ins. maybe the bun team refers to global objects as a built-ins as well? who knows anymore

**Comment by pick7** [comment]
Based on the current docs and the behavior in your example, `--drop` is matching **property-access/global-style call expressions**, not arbitrary local bindings.

That explains this result:

```ts
test()      // local binding, kept
_x.test()   // local object binding, kept
x.test()    // global/property access, dropped when --drop=x
```

The docs example is `--drop=console`, which is a global object/property family. The API reference also describes it as dropping calls to matching property accesses. Dropping a local function named `test` would be much harder to do safely because local identifiers can be shadowed, imported, reassigned, or have call arguments with side effects.

For custom debug-only code, I would use a compile-time flag instead:

```ts
if (DEBUG) {
  test()
}
```

and build with:

```sh
bun build index.ts --define DEBUG=false --minify
```

or put debug helpers behind an object you only use for that purpose:

```ts
globalThis.__debug?.test()
```

then drop that property path if it matches your bundle shape.

If the desired behavior is “remove calls to this local function and then tree-shake the function declaration,” that is more of a feature request for Bun's bundler than something `--drop` appears to support today. The docs could probably be clearer by saying “global/property call paths such as `console` or `foo.bar`,” not local function declarations.
